### PR TITLE
feat(loops): refactor scratchpad configuration to support per-hat

### LINKED ISSUE #174: [Feature]: Add configuration option to disable scratchpad injection and instructions
### Use case

Currently, the scratchpad (`.ralph/agent/scratchpad.md`) is unconditionally injected into every agent's prompt on every iteration, and the `### 0b. SCRATCHPAD` instructions section telling agents to write to it is always present. There is no way to disable this behavior.

This is problematic in multi-hat configurations for several scenarios:

### 1. Adversarial HAT configurations

In setups where HATs are intentionally adversarial (e.g., a Builder and a Confessor that evaluates the Builder's honesty), the Confessor can read the Builder's reasoning in the scratchpad, which defeats the purpose of an independent evaluation. The scratchpad leaks the internal reasoning of one agent to another, potentially biasing their output.

### 2. Phase-separated workflows with independent state

When HATs are designed to operate in distinct phases (e.g., Design → Implementation → Review), each phase may manage its own structured files in `.ralph/agent/` for state persistence. In these cases:

- The design HAT might write to `.ralph/agent/design-notes.md`
- The implementation HAT might write to `.ralph/agent/implementation-log.md`
- The review HAT might write to `.ralph/agent/review-findings.md`

The shared scratchpad becomes noise — polluting each phase with irrelevant context from previous phases. Once the design phase is finished and implementation begins, there's no need to read what the design HATs wrote in the scratchpad, especially if the relevant output is already captured in purpose-built files.

### 3. Token efficiency

The scratchpad budget is 16,000 characters (~4,000 tokens) of content injected into every prompt. For HATs that don't need cross-iteration continuity via the scratchpad (because they use other mechanisms), this is wasted context window.

### Proposed solution

Add a configuration option to `core` in `ralph.yml` to disable the scratchpad:

```yaml
core:
  scratchpad: ".ralph/agent/scratchpad.md"
  scratchpad_inject: false  # default: true
```

When `scratchpad_inject` is set to `false`:

1. **Skip scratchpad file injection** — `prepend_scratchpad()` in `EventLoop` should return the prompt unchanged.
2. **Skip the `### 0b. SCRATCHPAD` instructions section** — `core_prompt()` in `HatlessRalph` should omit the section that tells agents to write to the scratchpad. Injecting instructions to write to a file that won't be read back makes no sense and wastes tokens.
3. **Skip scratchpad references in `### STATE MANAGEMENT`** — The state management guidance should not reference the scratchpad when it's disabled.
4. **Skip scratchpad references in `## WORKFLOW`** — Workflow steps like "update scratchpad" should be omitted or adapted.

The `## ORIENTATION` section, guardrails, and other core prompt sections should remain unaffected.

### Area

Other

### Backend(s) impacted

_No response_

### Alternatives considered

_No response_

### Additional context

In solo mode (no hats defined), the scratchpad is the **primary continuity mechanism** between iterations. Since each iteration starts with completely fresh context and there are no alternative state channels (events are routing signals, not narrative state), disabling the scratchpad in solo mode would effectively break the agent's ability to maintain state. 

The implementation should either:
- **Ignore `scratchpad_inject: false` in solo mode** and log a warning, or
- **Document clearly** that this option is intended for multi-hat configurations where hats manage their own state files

### PR DESCRIPTION
**Pull request by luisherranz**
## Summary

Closes https://github.com/mikeyobrien/ralph-orchestrator/issues/174.

Refactors the scratchpad configuration from a plain string path (`core.scratchpad: "..."`) into a structured `ScratchpadConfig` type with `enabled` and `path` fields, and adds per-hat scratchpad overrides so each hat can use a custom scratchpad path, disable scratchpad entirely, or inherit the global setting.

- Introduces `ScratchpadConfig` struct with custom serde deserializers supporting both legacy plain-string and structured object YAML formats (backwards-compatible)
- Adds optional `scratchpad` field to `HatConfig` with resolution order: hat override > `core.scratchpad` > defaults
- Updates prompt generation in `HatlessRalph` to conditionally suppress all scratchpad-related sections (ORIENTATION, SCRATCHPAD, STATE MANAGEMENT, WORKFLOW, EVENT WRITING) when scratchpad is disabled
- Updates `EventLoop` to resolve per-hat scratchpad config before each iteration, skip scratchpad injection/verification when disabled, and fall back to global scratchpad for guidance persistence
- Adds solo-mode safety guard that force-enables scratchpad when no hats are defined (scratchpad is the only continuity mechanism in solo mode)
- Updates documentation for configuration, architecture, agents, and memories-and-tasks guides

## Test plan

- [ ] `cargo test -p ralph-core` passes (includes 20+ new tests for scratchpad config parsing, resolution, prompt generation, and edge cases)
- [ ] `cargo test` passes across all crates
- [ ] Verify legacy `core.scratchpad: "path"` string format still works (backwards-compatible deserialization)
- [ ] Verify structured `core.scratchpad: { enabled: true, path: "..." }` format works
- [ ] Verify hat-level `scratchpad` overrides (plain string, structured, disabled, absent/inherit)
- [ ] Verify prompts omit all scratchpad references when `enabled: false`
- [ ] Verify prompts use custom path when hat specifies one
- [ ] Verify solo mode forces scratchpad enabled when no hats defined

### CODE REVIEW

File: crates/ralph-core/src/event_loop/mod.rs [hunk 1]
```diff
@@ -365,12 +383,21 @@ impl EventLoop {
...
     }
 
-    /// Returns the scratchpad path based on loop context or config.
+    /// Returns the scratchpad path based on loop context or active scratchpad config.
     fn scratchpad_path(&self) -> PathBuf {
         self.loop_context
             .as_ref()
             .map(|ctx| ctx.scratchpad_path())
-            .unwrap_or_else(|| PathBuf::from(&self.config.core.scratchpad))
+            .unwrap_or_else(|| PathBuf::from(&self.ralph.active_scratchpad().path))
```
Comment (mikeyobrien) [inline crates/ralph-core/src/event_loop/mod.rs]: This still ignores per-hat `scratchpad.path` whenever `loop_context` is present (the normal CLI path via `EventLoop::with_context`). In practice that means custom per-hat scratchpad files are never used for file IO during real runs. Could we resolve from `active_scratchpad.path` first, then only apply loop-context remapping for the default/global path?

File: crates/ralph-core/src/hatless_ralph.rs [hunk 2]
```diff
@@ -332,8 +350,13 @@ You MUST NOT get distracted by workflow mechanics — they serve this goal.
             return false;
         }
 
+        // When scratchpad is disabled, no scratchpad to check — treat as fresh start
+        if !self.active_scratchpad.enabled {
```
Comment (mikeyobrien) [inline crates/ralph-core/src/hatless_ralph.rs]: Returning `true` whenever scratchpad is disabled makes FAST PATH trigger on every activation when `starting_event` is set (not just the first run). That can cause repeated re-publication of the starting event instead of handling the current triggered event. Suggest gating FAST PATH on an actual fresh-start signal (e.g., first iteration or `task.start`/`task.resume` context), not on `scratchpad.enabled == false`.

File: crates/ralph-core/src/event_loop/mod.rs [hunk 3]
```diff
@@ -753,6 +784,21 @@ impl EventLoop {
                 let active_hat_ids = self.determine_active_hat_ids(&regular_events);
                 self.record_hat_activations(&active_hat_ids);
                 self.state.last_active_hat_ids = active_hat_ids.clone();
+
+                // Resolve scratchpad config for the active hat (or global default)
+                // Must happen before determine_active_hats() which borrows self immutably
+                let resolved_scratchpad = if let Some(hat_id) = active_hat_ids.first() {
```
Comment (mikeyobrien) [inline crates/ralph-core/src/event_loop/mod.rs:1015]: This resolution happens *after* `update_robot_guidance()` above. `persist_guidance_to_scratchpad()` uses `active_scratchpad`, so guidance can be persisted to the previous hat's scratchpad/path for one iteration. Can we move active scratchpad resolution before guidance persistence/injection?

**Review by mikeyobrien — CHANGES_REQUESTED** [review]
Thanks for the refactor—this is close, but I found a few correctness issues that need to be fixed before merge.

File: crates/ralph-core/src/event_loop/mod.rs (hunk 1 above)
Comment (luisherranz) [reply crates/ralph-core/src/event_loop/mod.rs]: Fixed in https://github.com/mikeyobrien/ralph-orchestrator/pull/186/commits/abb75e4c6619ffa1fc1648a38e2062470b6aebc8.

`scratchpad_path()` now checks if the active path differs from the global default. Custom per-hat paths are resolved relative to the workspace; the default path still uses `ctx.scratchpad_path()`.

File: crates/ralph-core/src/hatless_ralph.rs (hunk 2 above)
Comment (luisherranz) [reply crates/ralph-core/src/hatless_ralph.rs]: Fixed in https://github.com/mikeyobrien/ralph-orchestrator/commit/abb75e4c6619ffa1fc1648a38e2062470b6aebc8.

Added an `iteration` field to `HatlessRalph` set by `EventLoop`. FAST PATH now only triggers on iteration 0.

File: crates/ralph-core/src/event_loop/mod.rs (hunk 3 above)
Comment (luisherranz) [reply crates/ralph-core/src/event_loop/mod.rs:1015]: Fixed in https://github.com/mikeyobrien/ralph-orchestrator/commit/abb75e4c6619ffa1fc1648a38e2062470b6aebc8.

Moved scratchpad resolution + `set_active_scratchpad` before `update_robot_guidance` in both the solo and multi-hat paths.

**Comment by luisherranz** [comment]
I've made those changes, but just so you know, I had the PR in draft because I haven't had time to test it yet. I hope to be able to test it in the next few days.

**Comment by mikeyobrien** [comment]
> I've made those changes, but just so you know, I had the PR in draft because I haven't had time to test it yet. I hope to be able to test it in the next few days.

@luisherranz, no worries - looking forward to get this out and thank you for working on it!

**Comment by luisherranz** [comment]
Still testing. So far so good but I want to test this thoroughly first to make sure that this API works well and is flexible. Also, I want to clean up the PR before marking it as ready for review because I've been merging a lot of branches with bug fixes.

**Comment by matbgn** [comment]
Looks very sound! Thank you for your work @luisherranz :pray: Eager to see it merged.

**Comment by luisherranz** [comment]
I think this is finally ready for review. I've tested it extensively, and I think the new API serves its purpose.

---

This was a good read, by the way, and I think it aligns with the "avoid sharing reasoning among opposing agents" strategy.

https://www.anthropic.com/engineering/harness-design-long-running-apps

**Comment by luisherranz** [comment]
@mikeyobrien, these are some (AI-extracted) opportunities to use this new functionality in the existing presets, divided by the three types I've been using in my experiments. No need to change current presets if they work well, but I just wanted to bring this to your attention to see what you thought.

---

## Preset opportunities for per-hat scratchpad

This PR introduces per-hat scratchpad configuration. Below are presets that could benefit from it, grouped by the type of improvement.

### Adversarial gates / validators — `scratchpad: { enabled: false }`

These hats are designed to challenge another hat's output with fresh eyes. Sharing a scratchpad undermines their independence by leaking the producer's reasoning and assumptions.

| Preset | Hat | Validates | Why disable |
|--------|-----|-----------|-------------|
| `code-assist.yml` | `critic` | `builder` | Explicitly called "Fresh-Eyes Critic" — instructions say "You are not the builder. That separation matters." |
| `code-assist.yml` | `finalizer` | all | Completion gate — should judge overall outcome independently |
| `pdd-to-code-assist.yml` | `design_critic` | `architect` | Adversarial design review gate — should not inherit the architect's reasoning |
| `pdd-to-code-assist.yml` | `critic` | `builder` | Same fresh-eyes concern as code-assist |
| `pdd-to-code-assist.yml` | `validator` | `builder` | "Exhaustive quality gate with manual E2E testing" — should judge the artifact, not the rationale |
| `debug.yml` | `tester` | `investigator` | Tests hypotheses — reading the investigator's reasoning risks confirmation bias |
| `debug.yml` | `verifier` | `fixer` | Verifies the fix solves the original problem — should not inherit the fixer's assumptions |
| `autoresearch.yml` | `judge` | `implementer` | Scores experiments against a rubric — instructions emphasize objectivity and consistency |
| `autoresearch.yml` | `evaluator` | `judge` + metrics | Keep/discard decision should rely on metrics and scores, not the strategist's optimism |
| `review.yml` | `analyzer` | `reviewer` | "Second adversarial pass" — anchoring on the reviewer's framing defeats the purpose |
| `review.yml` | `closer` | `analyzer` | Synthesizes and decides follow-up vs completion — should work from findings files only |
| `research.yml` | `synthesizer` | `researcher` | Reviews findings for contradictions — inheriting the researcher's framing may hide gaps |
| `wave-review.yml` | `synthesizer` | `reviewer` (wave) | Should aggregate from structured wave results, not any single reviewer's scratchpad |

### Stateless single-shot hats — `scratchpad: { enabled: false }`

These hats activate once per cycle, do one focused job, and emit. They don't benefit from cross-iteration continuity and the scratchpad prompt sections (ORIENTATION, SCRATCHPAD, STATE MANAGEMENT, WORKFLOW) add noise.

| Preset | Hat | Why disable |
|--------|-----|-------------|
| `autoresearch.yml` | `benchmarker` | Pure measurement — reads `autoresearch.md`, runs command, emits metrics |
| `autoresearch.yml` | `implementer` | Executes one plan and emits — no cross-iteration state needed |
| `pdd-to-code-assist.yml` | `committer` | Creates a commit and emits — purely mechanical |
| `pdd-to-code-assist.yml` | `task_writer` | Converts plan into task files — single transformation pass |
| `merge-loop.yml` | `cleaner` | Post-merge cleanup — single-shot |
| `merge-loop.yml` | `failure_handler` | Error recovery — single-shot |
| `wave-review.yml` | `reviewer` | Parallel wave workers — short-lived, get focused prompts |

### Producers / implementers — per-hat scratchpad path

These hats build state across iterations and would benefit from an isolated scratchpad path so their notes don't leak to other hats (especially the adversarial gates above).

| Preset | Hat | Suggested path | Why isolate |
|--------|-----|----------------|-------------|
| `code-assist.yml` | `builder` | `.ralph/agent/scratchpad-builder.md` | TDD cycle state (RED/GREEN/REFACTOR) should not leak to critic |
| `pdd-to-code-assist.yml` | `builder` | `.ralph/agent/scratchpad-builder.md` | Same TDD concern |
| `pdd-to-code-assist.yml` | `architect` | `.ralph/agent/scratchpad-architect.md` | Design reasoning should not anchor the design critic |
| `pdd-to-code-assist.yml` | `inquisitor` | `.ralph/agent/scratchpad-inquisitor.md` | Question strategy should not leak to the architect |
| `debug.yml` | `investigator` | `.ralph/agent/scratchpad-investigator.md` | Debugging hypotheses should not bias the hypothesis tester |
| `autoresearch.yml` | `strategist` | `.ralph/agent/scratchpad-strategist.md` | Experiment rationale should not bias the judge |
| `ralph.m.yml` | `explorer` | `.ralph/agent/scratchpad-explorer.md` | Market research state is distinct from builder/shipper state |
| `ralph.m.yml` | `builder` | `.ralph/agent/scratchpad-builder.md` | Implementation state should be isolated from verifier |
| `ralph.e2e.yml` | `planner` | `.ralph/agent/scratchpad-planner.md` | Planning notes should not leak to builder or validator |
| `merge-loop.yml` | `merger` | `.ralph/agent/scratchpad-merger.md` | Merge state tracking across conflict resolution iterations |

**Comment by mikeyobrien** [comment]
@luisherranz big thanks for this one. The per-hat scratchpad work is thoughtful, the testing/docs pass looks strong, and I appreciate how much care you put into the API shape here.

I’m merging this now and splitting the remaining custom-hat prompt plumbing into a follow-up so this doesn’t sit any longer: #293
